### DISCUSSION TITLE
Finetuning Citrinet with Kenlm Model

### DISCUSSION THREAD

**Discussion by xaxaks** [NVIDIA-NeMo/NeMo#4065] [Q&A] [answered]
Hi all, 
Im currently finetuning the stt_en_citrinet_1024 model and managed to obtain a WER of 12.92 percent using this configuration
![image](https://user-images.githubusercontent.com/76435723/165267511-a496d6d4-8313-4c59-8c2b-7ee5580d76cf.png)
![image](https://user-images.githubusercontent.com/76435723/165267549-756fcace-d0a7-4e6a-a9d4-2899ef1c54e6.png)
Tensorboard graph
![image](https://user-images.githubusercontent.com/76435723/165265483-e9237615-8286-4e36-b49f-39e1424a0e09.png)
I tried to finetune and lower certain parameters (batch size, min_lr, lr), but the WER seems to fluctuate and worsen. 
After training and evaluating the kenlm model, im able to obtain these results 
![image](https://user-images.githubusercontent.com/76435723/165266198-61b1e6ea-835a-4aac-8eae-1ca49736b09f.png)
I have a few queries : 
- what does the oracle WER/CER mean ? does it stand for the best case scenario ?
- are there any parameters that that i should be finetuning as well besides batch size, min_lr and lr ? 
- looking through the forum, will finetuning a STT En Citrinet 1024 Gamma 0.25 pretrained model give me a better WER ?

My aim is to get below 10 percent WER and deploy this on Riva and use it for real-time ASR streaming.

**Comment by titu1994** [comment, marked as answer]
Oracle wer uses the ground truth text at test time (which is usually never the case in real world) to compute whether there existed any beam candidate which best minimized the edit distance to this test time ground truth. 

It is a oracle measure stating what would have been the performance if a perfect LM was used to perfectly create and rank the beams. It is only used to measure the lower bound of the asr models accuracy in terms of wer, given a perfect language model. 

Depending on the size of the dataset, you may want to reduce your spec augment to speed up convergence. You can also load the decoder weights as shown in the tutorial if the shapes check out to speed up convergence and get better results usually. Other than that, there are only so many parameters that can affect training accuracy. You could try using a RNNT model such as ContextNet or Conformer Transducers if all you want is Nemo offline accuracy (and not use in NVIDIA Riva) which should dramatically improve wer and convergence speed. Though we don't yet support external LM beam search with RNNT

Gamma Citrinet may do better yes, it's more updated and usually gets close to same numbers as original Citrinet. Also Riva deployment requires gamma Citrinets for streaming mode anyway
